File: 0023. Merge k Sorted Lists/s1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
	ListNode* mergeKLists(vector<ListNode*>& lists) {
		ListNode* ans = new ListNode();
		ListNode* dummy = ans;
		auto comparer = [](ListNode* a, ListNode* b) {
			return a->val > b->val;
		};
		priority_queue<ListNode*, vector<ListNode*>, decltype(comparer)> pq(
			comparer);
		int n = lists.size();
		for (int i = 0; i < n; i++) {
			if (lists[i]) pq.push(lists[i]);
		}
		while (!pq.empty()) {
			ans->next = pq.top();
			pq.pop();
			if (ans->next->next) {
				pq.push(ans->next->next);
			}
			ans = ans->next;
		}
		return dummy->next;
	}
};

File: 0023. Merge k Sorted Lists/s1.java
/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
class Solution {
	public ListNode mergeKLists(ListNode[] lists) {
		if (lists.length == 0) return null;
		Comparator<ListNode> comparator = new ListNodeComparator();
		PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(lists.length, comparator);
		for (ListNode i : lists) {
			if (i != null) pq.add(i);
		}
		ListNode ans = new ListNode();
		ListNode dummy = ans;
		while (!pq.isEmpty()) {
			ans.next = pq.peek();
			pq.remove();
			if (ans.next.next != null) {
				pq.add(ans.next.next);
			}
			ans = ans.next;
		}
		return dummy.next;
	}
}

public class ListNodeComparator implements Comparator<ListNode> {
	@Override
	public int compare(ListNode a, ListNode b) {
		if (a.val < b.val) return -1;
		return 1;
	}
}